bin/user/purpleair.py

```diff
@@ -204,21 +210,28 @@ def get_and_update_missed(key):
 
     if missed:
         loginf("sensor didn't report field(s): %s" % ','.join(missed))
-
-    # for each concentration counter grab the average of the A and B channels and push into the record
-    for key in ['pm1_0_cf_1', 'pm1_0_atm', 'pm2_5_cf_1', 'pm2_5_atm', 'pm10_0_cf_1', 'pm10_0_atm']:
-        if is_data_from_purpleair_website:
-            valA = float(j[key])
-            valB = float(k[key])
-        else:
-            valA = float(j[key])
-            valB = float(j[key + '_b'])
-        if valA == 0.0 and valB != 0.0:
-            record[key] = valB
-        elif valB == 0.0 and valA != 0.0:
-            record[key] = valA
-        else:
-            record[key] = (valA + valB) / 2.0
+        
+    #when Lastseen field is older than 10 minutes do not return any particles data
+    if (int(time.time()) - int(LastSeen) < valid_timeout):
+        # for each concentration counter grab the average of the A and B channels and push into the record
+        
+        # NEWLY are values from PA website json with dot so it´s necessary to remap it
+        remap_dot = {'pm1_0_cf_1':'pm1.0_cf_1','pm1_0_atm':'pm1.0_atm','pm2_5_cf_1':'pm2.5_cf_1',\
+                   'pm2_5_atm':'pm2.5_atm','pm10_0_cf_1':'pm10.0_cf_1','pm10_0_atm':'pm10.0_atm'}
+
+        for key in ['pm1_0_cf_1', 'pm1_0_atm', 'pm2_5_cf_1', 'pm2_5_atm', 'pm10_0_cf_1', 'pm10_0_atm']:
+            if is_data_from_purpleair_website:
+                valA = float(j[remap_dot[key] + '_a'])
+                valB = float(j[remap_dot[key] + '_b'])
+            else:
+                valA = float(j[key])
+                valB = float(j[key + '_b'])
+            if valA == 0.0 and valB != 0.0:
+                record[key] = valB
+            elif valB == 0.0 and valA != 0.0:
+                record[key] = valA
+            else:
+                record[key] = (valA + valB) / 2.0
     return record
```

Comment:
@xslima00 could you help me understand the rationale for returning a record here in the case that the data is more than 10 minutes old?

Reply:
@bakerkj Hello, I have my weewx instance running in google cloud. My weather station, router and 5G modem are plugged to backup power supply, so in case of blackout I am still sending data into cloud for next 4-5 hours. PurpleAir device has no backup power. During each archive interval is request created to PurpleAir website and response is latest data. In my case I was receiving for next 4h still same JSON as PA device didn´t send anything newer. This filled DB with "incorect" data. In JSON returned from PA API is field ```last_seen``` which represents time when data were last seen from device. In case ```current_time - last_seen_time``` difference is smaller than 600 seconds, I know my data are still OK. If not, I know for some reason my device didn´t upload anything and I don´t need to store it into DB (I know I still store temperature, etc. but who is using weewx on 99% will never use PA temperature but temperature from own device). When you check local device in your network I set ```LastSeen = j['response_date']``` so data are always valid.

Reply:
Thanks!